src/main/java/uk/gov/ida/matchingserviceadapter/services/VerifyAssertionService.java

```diff
@@ -0,0 +1,104 @@
+import uk.gov.ida.saml.security.SamlAssertionsSignatureValidator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static uk.gov.ida.saml.core.validation.errors.GenericHubProfileValidationSpecification.MISMATCHED_ISSUERS;
+import static uk.gov.ida.saml.core.validation.errors.GenericHubProfileValidationSpecification.MISMATCHED_PIDS;
+
+public class VerifyAssertionService extends AssertionService {
+
+    private String hubEntityId;
+    private static final String AUTHN_ASSERTION = "authn-assertion";
+    private static final String MDS_ASSERTION = "mds-assertion";
+    private static final String CYCLE_3_ASSERTION = "cycle-3-assertion";
+    private final VerifyMatchingDatasetUnmarshaller matchingDatasetUnmarshaller;
+    private AuthnContextFactory authnContextFactory = new AuthnContextFactory();
+
+    public VerifyAssertionService(InstantValidator instantValidator,
+                                  SubjectValidator subjectValidator,
+                                  ConditionsValidator conditionsValidator,
+                                  SamlAssertionsSignatureValidator hubSignatureValidator,
+                                  Cycle3DatasetFactory cycle3DatasetFactory,
+                                  String hubEntityId,
+                                  VerifyMatchingDatasetUnmarshaller matchingDatasetUnmarshaller) {
+        super(instantValidator, subjectValidator, conditionsValidator, hubSignatureValidator, cycle3DatasetFactory);
+        this.hubEntityId = hubEntityId;
+        this.matchingDatasetUnmarshaller = matchingDatasetUnmarshaller;
+    }
+
+    @Override
+    public void validate(String requestId, List<Assertion> assertions) {
+        Map<String, List<Assertion>> assertionMap = assertions.stream()
+                .collect(Collectors.groupingBy(this::classifyAssertion));
+
+        List<Assertion> authnAssertions = assertionMap.get(AUTHN_ASSERTION);
+        if (authnAssertions.isEmpty() || authnAssertions.size() != 1) {
+            throw new SamlResponseValidationException("Exactly one authn statement is expected.");
+        }
+
+        List<Assertion> mdsAssertions = assertionMap.get(MDS_ASSERTION);
+        if (mdsAssertions.isEmpty() || mdsAssertions.size() != 1) {
+            throw new SamlResponseValidationException("Exactly one matching dataset assertion is expected.");
+        }
+
+        authnAssertions.forEach(a -> validateHubAssertion(a, requestId, hubEntityId, IDPSSODescriptor.DEFAULT_ELEMENT_NAME));
+        mdsAssertions.forEach(a -> validateHubAssertion(a, requestId, hubEntityId, IDPSSODescriptor.DEFAULT_ELEMENT_NAME));
+        assertionMap.getOrDefault(CYCLE_3_ASSERTION, emptyList()).forEach(a -> validateCycle3Assertion(a, requestId, hubEntityId));
+
+        Assertion mdsAssertion = mdsAssertions.get(0);
+        Assertion authnAssertion = authnAssertions.get(0);
```

Comment:
If you did these two `.get(0)`s before validating the assertion you could avoid the `.forEach`s above - if you know there's only going to be one there's no need to iterate through them.